diff/diff.go

```diff
@@ -0,0 +1,293 @@
+	return Chunk{
+		Path:    path,
+		Deleted: []interface{}{A},
+	}
+}
+
+// a value has changed, but don't report the before or after
+type OpaqueChunk struct {
+	Path string
+}
+
+func (c OpaqueChunk) AtPath() string {
+	return c.Path
+}
+
+// Objects are considered unique by {Namespace, Kind, Name}
+type ObjectID struct {
+	Namespace string
+	Kind      string
+	Name      string
+}
+
+func (id ObjectID) String() string {
+	if id.Namespace == "" {
+		return id.Kind + " " + id.Name
+	}
+	return fmt.Sprintf("%s %s/%s", id.Kind, id.Namespace, id.Name)
+}
+
+type Object interface {
+	ID() ObjectID
+	Source() string
+}
+
+// ObjectSet is a set of several objects which can be diffed
+// collectively.
+type ObjectSet struct {
+	Source  string
+	Objects map[ObjectID]Object
+}
+
+func MakeObjectSet(source string) *ObjectSet {
+	return &ObjectSet{
+		Source:  source,
+		Objects: map[ObjectID]Object{},
+	}
+}
+
+type ObjectSetDiff struct {
+	A, B      *ObjectSet
+	OnlyA     []Object
+	OnlyB     []Object
+	Different map[ObjectID][]Difference
+}
+
+func MakeObjectSetDiff(a, b *ObjectSet) ObjectSetDiff {
+	return ObjectSetDiff{
+		A:         a,
+		B:         b,
+		Different: map[ObjectID][]Difference{},
+	}
+}
+
+// Diff calculates the differences between one model and another
+func DiffSet(a, b *ObjectSet) (ObjectSetDiff, error) {
+	diff := MakeObjectSetDiff(a, b)
+
+	// A - B and A ^ B at the same time
+	for id, objA := range a.Objects {
+		if objB, found := b.Objects[id]; found {
+			objDiff, err := DiffObject(objA, objB)
+			if err != nil {
+				return diff, err
+			}
+			if len(objDiff) > 0 {
+				diff.Different[id] = objDiff
+			}
+		} else {
+			diff.OnlyA = append(diff.OnlyA, objA)
+		}
+	}
+	// now, B - A
+	for id, objB := range b.Objects {
+		if _, found := a.Objects[id]; !found {
+			diff.OnlyB = append(diff.OnlyB, objB)
+		}
+	}
+	return diff, nil
+}
+
+type Differ interface {
+	Diff(a Differ, path string) ([]Difference, error)
+}
+
+var ErrNotDiffable = errors.New("values are not diffable")
+
+// Diff one object with another. This assumes that the objects being
+// compared are supposed to represent the same logical object, i.e.,
+// they were identified with the same ID. An error indicates they are
+// not comparable.
+func DiffObject(a, b Object) ([]Difference, error) {
+	if a.ID() != b.ID() {
+		return nil, errors.New("objects being compared do not have the same ID")
+	}
+
+	// Special case at the top: if these have different runtime types,
+	// they are not comparable.
+	typA, typB := reflect.TypeOf(a), reflect.TypeOf(b)
+	if typA != typB {
+		return nil, ErrNotDiffable
+	}
+	return diffValue(reflect.ValueOf(a), reflect.ValueOf(b), typA, "")
+}
+
+var differType = reflect.TypeOf((*Differ)(nil)).Elem()
+
+// Compare two reflected values and compile a list of differences
+// between them.
+func diffValue(a, b reflect.Value, typ reflect.Type, path string) ([]Difference, error) {
+	if typ.Implements(differType) {
```

Comment:
Could all the diff*'s implement the differType? Then you wouldn't need the switch below?